default/python/AI/ResourcesAI.py

```diff
@@ -148,115 +187,338 @@ def check_avoid_change_due_to_bureaucracy(self, pinfo: PlanetFocusInfo, focus: s
+        target_stability = planet.currentMeterValue(fo.meterType.targetHappiness) if stability is None else stability
+        if self.have_liberty and focus == RESEARCH:
+            research_target += PlanetUtilsAI.adjust_liberty(planet, planet.currentMeterValue(fo.meterType.population))
+        if self.have_automation:
+            min_stability = get_named_real("PRO_ADAPTIVE_AUTO_MIN_STABILITY")
+            flat = get_named_real("PRO_ADAPTIVE_AUTO_TARGET_INDUSTRY_FLAT")
+            industry_target += flat * adjust_direction(min_stability, current_stability, target_stability)
+        if self.have_ai:
+            min_stability = get_named_real("LRN_NASCENT_AI_MIN_STABILITY")
+            flat = get_named_real("LRN_NASCENT_AI_TARGET_RESEARCH_FLAT")
+            research_target += flat * adjust_direction(min_stability, current_stability, target_stability)
+        if target_stability < 0:
+            industry_target = research_target = 0.0
+        # There are a lot more adjustments, some also depend on supply connection. Hopefully one day we get
+        # meter_update with adjusted stability...
+
+        return Output(
+            industry=industry_target,
+            research=research_target,
+            influence=influence_target,
+            stability=target_stability,
+            rating=self.production_rating(industry_target, research_target, influence_target, target_stability, focus),
         )
 
-    def calculate_planet_infos(self, pids):  # noqa: max-complexity
-        """Calculates for each possible focus the target output of each planet and stores it in planet info
-        It excludes baked planets from consideration.
-        Note: The results will not be strictly correct if any planets have global effects
+    def calculate_planet_infos(self):  # noqa complexity
         """
-        # TODO this function depends on the specific rule that off-focus meter value are always the minimum value
+        Calculates for each possible focus an estimated target output of each planet and stores it in planet info.
+        """
+
+        def set_focus(focus_):
+            if focus_ in pinfo.planet.availableFoci and pinfo.planet.focus != focus_:
+                fo.issueChangeFocusOrder(pid, focus_)
+
         universe = fo.getUniverse()
-        unbaked_pids = [pid for pid in pids if pid not in self.baked_planet_info]
-        planet_infos = [(pid, self.all_planet_info[pid], self.all_planet_info[pid].planet) for pid in unbaked_pids]
-        for pid, pinfo, planet in planet_infos:
-            if INDUSTRY in planet.availableFoci and planet.focus != INDUSTRY:
-                fo.issueChangeFocusOrder(pid, INDUSTRY)  # may not be able to take, but try
-
-        def merge(po1, po2):
-            """merge two partially determined possible_outputs"""
-            return (
-                po1[0] or po2[0],
-                po1[1] or po2[1],
-                po1[2] or po2[2],
+        for pid, pinfo in self.planet_info.items():
```

Comment:
This code is very complex, each cycle is coupled with the next one.

Let's do some generalization with a double iteration on each step. Readability > performance. 

```python
def collect_setimates(planet_info, focus, callback):
    for pid, pinfo in planet_info.items():
        set_focus(pid, focus)
    universe.updateMeterEstimates(list(planet_info))
    for pid, pinfo in planet_info.items():
        pinfo.possible_output[focus] = callback(pinfo.planet, focus)
```